### PR TITLE
Feature/v1-208 | New course validation rules.

### PR DESCRIPTION
**Pull request by dapodev**
# Changes

Validation rules changed on `title`, `description`, `materials` and `test` fields.
Rules described here:

Title of the course:
- the user cannot save only special characters
- the user cannot save only numbers
- the user cannot save only spaces
- if the user enters a title with extra spaces, these extra spaces should be removed after saving
- the minimum length should be 2 characters
- the maximum length should be 100 characters
- after saving, the first letter of the name becomes uppercase

Description of the course:
- the user cannot save only special characters
- the user cannot save only numbers
- if the user enters a description with extra spaces, these extra spaces should be removed after saving
- the minimum length should be 50 characters
- the maximum length should be 3000 characters
- after saving, the first letter of the name becomes uppercase

Text material:
- the user cannot save only special characters
- the user cannot save only numbers
- the user cannot save only spaces
- if the user enters a title with extra spaces, these extra spaces should be removed after saving
- the minimum length should be 10 characters
- the maximum length should be 5000 characters
- after saving, the first letter of the name becomes uppercase

Questions for the test:
- the minimum amount of the questions in the test should be 5
- the user cannot save only special characters
- the user cannot save only numbers
- the user cannot save only spaces
- if the user enters a title with extra spaces, these extra spaces should be removed after saving
- the minimum length should be 10 characters
- the maximum length should be 1000 characters
- after saving, the first letter of the name becomes uppercase

Aswers on the questions:
- the user cannot save only spaces
- if the user enters a title with extra spaces, these extra spaces should be removed after saving

UPD:
- rewrote code with `yup` npm package
- added complexity update feature
- added test title update feature
- synced bodies of both update and create endpoints

### CODE REVIEW

**Review by SaXaPhonist — COMMENTED** [review]
maybe it will be better to use special library for validation, for exmp "Ajv"

**Comment by dapodev** [comment]
> maybe it will be better to use special library for validation, for exmp "Ajv"

Why didn't you propose replace all the validation just before hundreds of lines of code were written during edit and create course previours PRs? 😆

**Comment by SaXaPhonist** [comment]
> 

not in all cases of course  your functions are still important for unique cases

**Comment by dapodev** [comment]
Yep, you're right. Starting working on refactoring 😄

File: backend/src/controllers/admin/editCourse.ts [hunk 1]
```diff
@@ -3,15 +3,20 @@ import { NextFunction, Request, Response } from 'express';
...
+  isValidComplexity,
+  validateMaterials,
+  validateTechnologies,
+  validateTest,
+  validateTitle,
+} from 'utils/validation/courses';
+import validateDescription from 'utils/validation/courses/validateDescription';
+import { convertToTypeUnsafe } from 'utils/typeConversion/common';
 
 const editCourse = async (
```
[the remark this answers is not in the document]
Comment (dapodev) [reply backend/src/controllers/admin/editCourse.ts:21]: yep, but why?

File: backend/src/db/providers/skillProvider.ts [hunk 2]
```diff
@@ -180,7 +180,7 @@ const skillsExist = async (ids?: string[] | ObjectId[]): Promise<boolean> => {
   return foundSkillsCount === ids.length;
 };
 
-const isProperTechnologies = async (techs: IUpdateCourseBody['skills']): Promise<boolean> => {
+const isProperTechnologies = async (techs: IUpdateCourseBody['technologies']): Promise<boolean> => {
```
Comment (SaXaPhonist) [inline backend/src/db/providers/skillProvider.ts:183]: why is it in providers? look like validation or utils

File: backend/src/interfaces/ICourses/IQueryCourses.ts [hunk 3]
```diff
@@ -32,17 +32,16 @@ interface IUpdateCourseBody {
   title?: ICourse['title'];
   avatar?: ICourse['avatar'];
   description?: ICourse['description'];
-  skills?: {
```
Comment (SaXaPhonist) [inline backend/src/interfaces/ICourses/IQueryCourses.ts:35]: ```Omit<ICourse, 'skills'>``` - is that what you need?

**Review by SaXaPhonist — APPROVED** [review]

File: backend/src/interfaces/ICourses/IQueryCourses.ts (hunk 3 above)
Comment (dapodev) [reply backend/src/interfaces/ICourses/IQueryCourses.ts:35]: Well, I could refactor this with `omit`, yeah. But we should specify here what actual fields we can update. So if we use `omit`, we will receive all the fields from course, even those we don't want to update

**Pull request merged**